### PR TITLE
[spi_device] Add initial batch of DIFs for flash / passthrough modes

### PR DESCRIPTION
**Pull request by a-will**
DIFs for handling the general command info table, passthrough filters,
address and data swapping, basic FIFO and memory interaction, flash
status registers, address modes, and identification are all contained in
this commit.

### CODE REVIEW

File: sw/device/lib/dif/dif_spi_device.c [hunk 1]
```diff
@@ -434,3 +480,539 @@ dif_result_t dif_spi_device_send(const dif_spi_device_t *spi,
...
+  if (status != kDifOk) {
+    return status;
+  }
+  if (offset >= (info.buffer_offset + info.buffer_len) ||
+      length > (info.buffer_offset + info.buffer_len - offset)) {
+    return kDifBadArg;
+  }
+  ptrdiff_t offset_from_base =
+      SPI_DEVICE_BUFFER_REG_OFFSET + info.buffer_offset + offset;
+  mmio_region_memcpy_from_mmio32(spi->dev.base_addr, offset_from_base, buf,
```
Comment (moidx) [inline sw/device/lib/dif/dif_spi_device.c:911]: Is `spi_memcpy` still needed?

**Review by moidx — APPROVED** [review]
I recommend to break big changes into smaller PRs in the future. For instance, you could have send a PR to change the device config/handle, and one or two more to add flash and passthrough modes.

File: sw/device/lib/dif/dif_spi_device.c (hunk 1 above)
Comment (a-will) [reply sw/device/lib/dif/dif_spi_device.c:911]: `spi_memcpy` is still needed: It handles the FIFO pointers for generic mode.

`dif_spi_device_read_flash_buffer()` is purely for the RAM-style access used in flash and passthrough modes (i.e. it's not FIFO-style).

File: sw/device/lib/dif/dif_spi_device.h [hunk 2]
```diff
@@ -170,26 +245,379 @@ dif_result_t dif_spi_device_tx_pending(const dif_spi_device_t *spi,
...
+  dif_spi_device_payload_io_t payload_io_type;
+  /**
+   * Whether to translate the address using the address swap mask and value
+   * provided to `dif_spi_device_passthrough_set_swap_address()`. Incompatible
+   * with a command with `address_type` that is
+   * `kDifSpiDeviceFlashAddrDisabled`.
+   */
+  bool passthrough_swap_address;
+  /** Whether the SPI host is receiving this command's payload phase.
+   * Incompatible with
```
[the remark this answers is not in the document]
Comment (a-will) [reply sw/device/lib/dif/dif_spi_device.h:399]: Huh, I must've messed up a rebase and lost them (and a few other edits). I'll fix that up.

File: sw/device/lib/dif/dif_spi_device.h [hunk 3]
```diff
@@ -170,26 +245,379 @@ dif_result_t dif_spi_device_tx_pending(const dif_spi_device_t *spi,
...
+  kDifSpiDeviceFlashBufferTypeMailbox,
+  /** SFDP region */
+  kDifSpiDeviceFlashBufferTypeSfdp,
+  /** Payload for uploaded commands */
+  kDifSpiDeviceFlashBufferTypePayload,
+  /** Count of buffer types */
+  kDifSpiDeviceFlashBufferTypes,
+} dif_spi_device_flash_buffer_type_t;
+
+// Configure cmd_info table
```
[the remark this answers is not in the document]
Comment (a-will) [reply sw/device/lib/dif/dif_spi_device.h]: Yup! Removed.

File: sw/device/lib/dif/dif_spi_device.h [hunk 4]
```diff
@@ -170,26 +245,379 @@ dif_result_t dif_spi_device_tx_pending(const dif_spi_device_t *spi,
...
+OT_WARN_UNUSED_RESULT
+dif_result_t dif_spi_device_set_flash_payload_swap(dif_spi_device_handle_t *spi,
+                                                   uint32_t mask,
+                                                   uint32_t replacement);
+
+/**
+ * Pop the first command from the uploaded command FIFO.
+ *
+ * @param spi A handle to a spi device.
+ * @param command A pointer to where the command can be stored.
```
[the remark this answers is not in the document]
Comment (a-will) [reply sw/device/lib/dif/dif_spi_device.h]: I sure didn't, haha. Fixed.

**Comment by a-will** [comment]
Looks like I broke generic mode somehow. Sorry I forgot I had an easy test for that already--Bootstrap is failing on the FPGA, haha. Debugging next!

**Pull request merged**